uber/models/department.py

```diff
@@ -0,0 +1,428 @@
+        order_by='Attendee.full_name',
+        viewonly=True)
+    members = relationship(
+        'Attendee',
+        backref=backref('assigned_depts', order_by='Department.name'),
+        cascade='save-update,merge,refresh-expire,expunge',
+        order_by='Attendee.full_name',
+        secondary='dept_membership')
+    memberships = relationship('DeptMembership', backref='department')
+    explicit_membership_requests = relationship(
+        'DeptMembershipRequest', backref='department')
+    requesting_attendees = relationship(
+        'Attendee',
+        backref=backref('requested_depts', order_by='Department.name'),
+        cascade='save-update,merge,refresh-expire,expunge',
+        primaryjoin='or_('
+                    'DeptMembershipRequest.department_id == Department.id, '
+                    'DeptMembershipRequest.department_id == None)',
+        secondary='dept_membership_request',
+        order_by='Attendee.full_name',
+        viewonly=True)
+    unassigend_requesting_attendees = relationship(
+        'Attendee',
+        cascade='save-update,merge,refresh-expire,expunge',
+        primaryjoin='and_(or_('
+                    'DeptMembershipRequest.department_id == Department.id, '
+                    'DeptMembershipRequest.department_id == None), '
+                    'not_(exists().where(and_('
+                    'DeptMembership.department_id == Department.id, '
+                    'DeptMembership.attendee_id == '
+                    'DeptMembershipRequest.attendee_id))))',
+        secondary='dept_membership_request',
+        order_by='Attendee.full_name',
+        viewonly=True)
+    parent = relationship(
+        'Department',
+        backref=backref(
+            'sub_depts',
+            order_by='Department.name',
+            cascade='all,delete-orphan'),
+        cascade='save-update,merge,refresh-expire,expunge',
+        remote_side='Department.id',
+        single_parent=True)
+
+    @hybrid_property
+    def member_count(self):
+        return len(self.memberships)
+
+    @member_count.expression
+    def member_count(cls):
+        return func.count(cls.memberships)
+
+    @classmethod
+    def to_id(cls, department):
+        if not department:
+            return None
+
+        if isinstance(department, six.string_types):
+            try:
+                department = int(department)
+            except ValueError:
+                return department
+
+        if isinstance(department, int):
+            # This is the same algorithm used by the migration script to
+            # convert c.JOB_LOCATIONS into department ids in the database.
+            prefix = '{:07x}'.format(department)
+            return prefix + str(uuid.uuid5(cls.NAMESPACE, str(department)))[7:]
+
+        return department.id
+
+    def checklist_item_for_slug(self, slug):
+        for item in self.dept_checklist_items:
+            if item.slug == slug:
+                return item
```

Comment:
Hm, won't this lead to some weird behavior? There can be multiple checklist items with the same slug, even in one department, because multiple attendees may complete the checklist. I thought that was the case, anyway.

Reply:
Ah, I just remembered, I added a unique constraint to `DeptChecklistItem` so there can only be one completed item of each type (slug) per department:
```
__table_args__ = (UniqueConstraint('department_id', 'slug'),)
```

Reply:
Oooh, I missed that. I... hm. I have NO IDEA if that will work long term or not. I'm going to bet 'yes', so let's roll with it!

Reply:
Oops, I feel bad for missing this, but in the past we have totally had multiple dept heads fill out the same checklist items.  You can see at https://super2017.uber.magfest.org/uber/dept_checklist/overview which depts had multiple people filling out most of the items:
- Arcade
-MIVS
- LAN
- Museum
- Rescuers
- Staff Suite
- CCG Tabletop
- Treasury
- Zombie Tag

With that being said, most of those steps were basically just "Yeah I looked at the shifts and they're fine" and such.  While some departments may decide they want multiple people independently completing that step, that's probably not important enough to hold up merging this IMO.

Reply:
I am aware of the multiple dept heads issue. For each slug, the migration keeps the first completed checklist item and discards the rest. That may be a bad assumption on my part.

Do we want multiple dept heads completing the same checklist item?

Reply:
> Do we want multiple dept heads completing the same checklist item?
￼
On balance I think we should omit that feature for now.

The more I think about it the more I realize that we fundamentally have two different types of checklist items:

1) Ones which are "go in and double check that what you have on this page looks good".  For example, "Yes, these shifts look good to me" and "Yes, I have reviewed the list of people in my department and don't notice anyone missing".  Those types of items are things we want multiple people to complete.

2) Ones which we expect only a single person to complete.  For example, "What signs are needed by your department" or "Please detail the amounts and denominations of cash and mpoints that you are requesting".  Most of the time, even departments where multiple people are filling out the checklist only have a single person complete that kind of item, because that's the person in charge of making those decisions.

So basically, I think it's best to keep one-person-per-checklist-item as the rule for now, because if we do change that in the future, it probably WON'T be to make every checklist item completable by every assigned person, but rather to add the ability to configure specific steps to be completable by multiple people, or something along those lines.  But that kind of thing is WAY out of scope for this PR, and I wouldn't want to make that decision without first seeing how dept heads use the new checklist system anyway.

tldr; better to keep it the way it is now.